packages/nuxt/src/app/composables/asyncData.ts

```diff
@@ -61,7 +61,7 @@ export interface AsyncDataOptions<
    * A `null` or `undefined` return value will trigger a fetch.
    * Default is `key => nuxt.isHydrating ? nuxt.payload.data[key] : nuxt.static.data[key]` which only caches data when payloadExtraction is enabled.
    */
-  getCachedData?: (key: string) => DataT
+  getCachedData?: <T = DataT>(key: string) => T
```

Comment:
Nice catch, there's no type constrain any more, I've applied you suggestion to return `any`, thanks!

Reply:
Ideally we'd _like_ it to constrain the return, I think.

I guess the issue is that the type of `nuxtApp.payload[key]` is `unknown` and ideally when using the function we should either assert that this type matches or perform some sanitisation to ensure it matches. (We might exempt `unknown` given that this is likely a common situation.)

Reply:
Agreed, I think constrain and type hint really matters.

I just found the type of data will be influenced by the return of `getCachedData`, for example

Here the type of `data` is `Ref<boolean | null>` 
```ts
const { data } = await useAsyncData("test2", async () => ({ hello: "world!" }), {
  getCachedData: (key) => {
    return true
  },
});
```

The cache `useNuxtApp().payload.data[key]` is `any`, which caused an `any` data. As of that, I think users should handle the return type of `getCachedData` if it's `any`.

Reply:
Wouldn't this be a great usecase for the new `NoInfer` type utility? This would make `getCachedData` not influence the type of `data`, but still give hints/constrain the type of `getCachedData`.
e.g.
```ts
type AsyncDataOptions<...> = {
    getCachedData?: (key: string) => NoInfer<DataT>;
}
```

As it's a rather new feature ([TS 5.4](https://devblogs.microsoft.com/typescript/announcing-typescript-5-4/#the-noinfer-utility-type)), we might reimplement it like so for now:
```ts
// see https://github.com/millsp/ts-toolbelt/blob/master/sources/Function/NoInfer.ts
type NoInfer<T> = [T][T extends any ? 0 : never];
```

Reply:
what a great idea!

Reply:
Thanks @94726, I've applied your suggestion, it looks good!